File: web/src/api/onboarding.ts
import type { AuthCache } from "../auth/storage";

export type CurrencyCode = "RUB" | "USD" | "EUR";
export type SnapshotType = "WEEKLY" | "MONTLY";

export type OnboardingInitialSnapshot = {
  incomes: Record<string, number>;
  expenses: Record<string, number>;
  savings_deposits: Record<string, number>;
  savings_withdrawals: Record<string, number>;
};

export type OnboardingState = {
  currency: CurrencyCode | null;
  snapshot_type: SnapshotType | null;
  income_categories: string[] | null;
  expense_categories: string[] | null;
  accounts: string[] | null;
  initial_capital: Record<string, number> | null;
  initial_snapshot: OnboardingInitialSnapshot | null;
};

export type OnboardingStatePatch = Partial<OnboardingState>;

export class OnboardingApiError extends Error {
  readonly status: number;
  readonly detail: unknown;

  constructor(status: number, message: string, detail?: unknown) {
    super(message);
    this.status = status;
    this.detail = detail;
    this.name = "OnboardingApiError";
  }
}

function authHeaders(token: string | null): HeadersInit {
  const base: Record<string, string> = { "Content-Type": "application/json" };
  if (token) base.Authorization = `Bearer ${token}`;
  return base;
}

async function parseError(response: Response): Promise<OnboardingApiError> {
  let detail: unknown = null;
  let message = `Ошибка сервера (${response.status})`;
  try {
    const data = (await response.json()) as { detail?: unknown };
    detail = data.detail;
    if (typeof detail === "string") {
      message = detail;
    } else if (detail && typeof detail === "object" && "message" in detail) {
      const m = (detail as { message?: unknown }).message;
      if (typeof m === "string") message = m;
    }
  } catch {
    /* keep default message */
  }
  return new OnboardingApiError(response.status, message, detail);
}

export async function getState(token: string | null): Promise<OnboardingState> {
  const response = await fetch("/api/v1/onboarding/state", {
    method: "GET",
    headers: authHeaders(token),
    credentials: "include",
  });
  if (!response.ok) throw await parseError(response);
  return (await response.json()) as OnboardingState;
}

export async function patchState(
  token: string | null,
  patch: OnboardingStatePatch,
): Promise<OnboardingState> {
  const response = await fetch("/api/v1/onboarding/state", {
    method: "PATCH",
    headers: authHeaders(token),
    credentials: "include",
    body: JSON.stringify(patch),
  });
  if (!response.ok) throw await parseError(response);
  return (await response.json()) as OnboardingState;
}

export async function complete(token: string | null): Promise<AuthCache> {
  const response = await fetch("/api/v1/onboarding/complete", {
    method: "POST",
    headers: authHeaders(token),
    credentials: "include",
  });
  if (!response.ok) throw await parseError(response);
  return (await response.json()) as AuthCache;
}

File: app/schemas/onboarding.py
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.user import UserResponse

CurrencyCode = Literal["RUB", "USD", "EUR"]
SnapshotType = Literal["WEEKLY", "MONTLY"]


class OnboardingInitialSnapshot(BaseModel):
    """Плановый снапшот первого периода, формируемый в онбординге.

    Ключ периода (``snapshot_key``) определяется на сервере по МСК времени
    и не передаётся клиентом.

    Attributes:
        incomes: Плановые доходы в разрезе категорий.
        expenses: Плановые расходы в разрезе категорий.
        savings_deposits: Плановые пополнения сбережений по счетам.
        savings_withdrawals: Плановые расходы сбережений по счетам.
    """

    incomes: dict[str, int] = Field(default_factory=dict)
    expenses: dict[str, int] = Field(default_factory=dict)
    savings_deposits: dict[str, int] = Field(default_factory=dict)
    savings_withdrawals: dict[str, int] = Field(default_factory=dict)


class OnboardingState(BaseModel):
    """Текущее состояние онбординга, хранимое в Redis.

    Каждое поле опционально — на разных шагах часть данных может быть
    не заполнена.

    Attributes:
        currency: Выбранная валюта.
        snapshot_type: Выбранный тип снапшота.
        income_categories: Список категорий доходов.
        expense_categories: Список категорий расходов.
        accounts: Список счетов (включая ``"Основной счёт"``).
        initial_capital: Стартовый капитал по счетам.
        initial_snapshot: Первый плановый снапшот.
    """

    currency: CurrencyCode | None = None
    snapshot_type: SnapshotType | None = None
    income_categories: list[str] | None = None
    expense_categories: list[str] | None = None
    accounts: list[str] | None = None
    initial_capital: dict[str, int] | None = None
    initial_snapshot: OnboardingInitialSnapshot | None = None


class OnboardingStatePatch(BaseModel):
    """Частичное обновление состояния онбординга.

    Содержит произвольное подмножество полей :class:`OnboardingState`,
    которые перезаписывают одноимённые ключи в Redis.
    """

    currency: CurrencyCode | None = None
    snapshot_type: SnapshotType | None = None
    income_categories: list[str] | None = None
    expense_categories: list[str] | None = None
    accounts: list[str] | None = None
    initial_capital: dict[str, int] | None = None
    initial_snapshot: OnboardingInitialSnapshot | None = None


class OnboardingCompleteResponse(BaseModel):
    """Ответ эндпоинта завершения онбординга.

    Attributes:
        access_token: Новый JWT с обновлёнными claim'ами.
        token_type: Тип токена, всегда ``bearer``.
        user: Учётная запись пользователя.
    """

    access_token: str
    token_type: str = "bearer"
    user: UserResponse
